regression/cbmc-library/float-nan-check/main.c

```diff
@@ -0,0 +1,66 @@
+#include <assert.h>
+#include <math.h>
+#include <stdint.h>
+
+// This test contains a set of operations performed on a floating point numbers
+// to verify --nan-check flag. The result of test may change when/if support
+// of NaN value propagation checking is supported. Current behaviour of
+// --nan-check flag is that it checks whether an arithmetic operation performed
+// on a floating point value(s) is generating new not-a-number value. If NaN is
+// generated because one of the values is already NaN the assertion is not
+// produced.
+
+union container {
+  uint32_t u;
+  float f;
+};
+
+int main(void)
+{
+  union container c = {UINT32_MAX};
+  int n;
+
+  // should generate assertion if --nan-check flag will support
+  // NaN propagation checking
+  float f = c.f;
+  f = c.f + 0.0;
+
```

Comment:
Rather than use constants (and so test just the constant propagation code), how about:

```
float my_NaN;
__CPROVER_assume(isnan(my_NaN));
float my_inf;
__CPROVER_assume(isinf(my_inf) && my_inf > 0.0);
float my_zero;
__CPROVER_assume(my_zero == 0.0f);
```

The results should be the same (well `my_zero` might be negative but I don't think it changes anything) but because they are non-deterministic, the whole solver will be run.

Reply:
Thanks for suggestions, I will have a play with non-deterministic approach.